File: src/multi/model/SubDerModel.java
package multi.model;

import ilog.concert.*;
import ilog.cplex.IloCplex;
import multi.data.InputData;
import multi.data.Parameter;
import multi.structure.Request;

public class SubDerModel extends BasePrimalModel {
	private double[] uValueDouble;
	public SubDerModel(InputData in, Parameter p) {
		super();
		this.in = in;
		this.p = p;
		this.ModelName = "SDP"+ "-R"+ in.getShipRouteSet().size() + "-T" + p.getTimeHorizon() + "-"+ FleetType + "-S" + randomSeed;
		this.uValueDouble = new double [p.getDemand().length];
		try{
			cplex = new IloCplex();

			publicSetting(cplex);
			frame();
		}catch (IloException e)
		{
			e.printStackTrace();
		}
	}


	protected void setDecisionVars() throws IloException
	{
		SetVesselDecisionVars();
		SetRequestDecisionVars();
	}

	protected void setObjectives() throws IloException
	{
		IloLinearNumExpr Obj=cplex.linearNumExpr();

		// item1 : Operating Cost
		// r∈R
		for(int r = 0; r<p.getShippingRouteSet().length; r++)
		{
			// h∈H
			for(int w = 0; w < p.getVesselPathSet().length; w++)
			{
				for (int h=0;h<p.getVesselSet().length;h++)
				{
					Obj.addTerm(p.getVesselTypeAndShipRoute()[h][r]
							*p.getShipRouteAndVesselPath()[r][w]
							*p.getVesselOperationCost()[h]
							, vVar[h][r]);
				}
			}
		}

		// i
		for(int i=0;i<p.getDemand().length;i++)
		{
			// item2 : Penalty Cost of unsatisfied Demand : c2i*Gi
			Obj.addTerm(p.getPenaltyCostForDemand()[i], gVar[i]);

			Request od = in.getRequestSet().get(i);
			// \phi_i
			for(int k=0;k<od.getNumberOfLadenPath();k++)
			{
				int j = od.getLadenPathIndexes()[k];
				// item3 : Demurrage of self-owned and leased containers and Rental cost on laden paths
				Obj.addTerm(p.getLadenPathCost()[j], xVar.get(i)[k]);
				Obj.addTerm(p.getLadenPathCost()[j], yVar.get(i)[k]);
				Obj.addTerm(p.getRentalCost()*p.getTravelTimeOnPath()[j], yVar.get(i)[k]);
			}

			// item4: Demurrage of self-owned containers for empty path
			// \theta
			for(int k=0;k<od.getNumberOfEmptyPath();k++)
			{
				int j = od.getEmptyPathIndexes()[k];

				Obj.addTerm(p.getEmptyPathCost()[j], zVar.get(i)[k]);
			}
		}

		cplex.addMinimize(Obj);
	}

	private IloRange[] C1;

	protected void setConstraints() throws IloException
	{
		setConstraint0();
		
		setConstraint1();
		setConstraint2();
		setConstraint3();
	}

	//(15)each ship route selected one vessel type
	private void setConstraint0() throws IloException
	{
		// r∈R
		for(int r = 0; r<p.getShippingRouteSet().length; r++)
		{
			IloLinearIntExpr left=cplex.linearIntExpr();

			// h∈H
			for(int h=0;h<p.getVesselSet().length;h++)
			{
				// r(h) == r
				left.addTerm(p.getVesselTypeAndShipRoute()[h][r], vVar[h][r]);
			}
			String ConstrName = "C0("+(r+1)+")";
			cplex.addEq(left,1, ConstrName);
		}
	}

	//(21)demand equation
	private void setConstraint1() throws IloException 
	{
		C1 = new IloRange[p.getDemand().length];

		//∀i∈I
		for(int i=0;i<p.getDemand().length;i++)
		{
			IloLinearNumExpr left=cplex.linearNumExpr();

			Request od = in.getRequestSet().get(i);
			//φ
			for(int k=0;k<od.getNumberOfLadenPath();k++)
			{
				left.addTerm(1, xVar.get(i)[k]);
				left.addTerm(1, yVar.get(i)[k]);
			}

			left.addTerm(1, gVar[i]);

			String ConstrName = "C1("+(i+1)+")";
			C1[i] = cplex.addEq(left ,p.getDemand()[i] + p.getMaximumDemandVariation()[i] * uValueDouble[i], ConstrName);
		}
	}

	//(22) Vessel Capacity Constraint
	private void setConstraint2() throws IloException
	{
		// ∀<n,n'>∈A'
		for(int nn = 0; nn<p.getTravelingArcsSet().length; nn++)
		{
			IloLinearNumExpr left=cplex.linearNumExpr();

			// i∈I
			for(int i=0;i<p.getDemand().length;i++)
			{
				Request od = in.getRequestSet().get(i);

				// φ
				for(int k=0; k<od.getNumberOfLadenPath(); k++)
				{
					int j = od.getLadenPathIndexes()[k];

					left.addTerm(p.getArcAndPath()[nn][j], xVar.get(i)[k]);
					left.addTerm(p.getArcAndPath()[nn][j], yVar.get(i)[k]);
				}

				//θ
				for(int k=0;k<od.getNumberOfEmptyPath();k++)
				{
					int j = od.getEmptyPathIndexes()[k];

					left.addTerm(p.getArcAndPath()[nn][j], zVar.get(i)[k]);
				}
			}

			// vessel capacity
			// r∈R
			for(int r = 0; r<p.getShippingRouteSet().length; r++)
			{
				// ω∈Ω
				for(int w=0;w<p.getVesselPathSet().length;w++)
				{
					// h∈H
					for(int h=0;h<p.getVesselSet().length;h++)
					{
						left.addTerm(-p.getArcAndVesselPath()[nn][w]
										*p.getShipRouteAndVesselPath()[r][w]
										*p.getVesselTypeAndShipRoute()[h][r]
										*p.getVesselCapacity()[h]
								, vVar[h][r]);
					}
				}
			}
			String ConstrName = "C3"+"("+(nn+1)+")";
			cplex.addLe(left,0, ConstrName);
		}
	}

	// (24)Containers flow conservation
	private void setConstraint3() throws IloException
	{
		// p \in P
		for(int pp = 0; pp < p.getPortSet().length; pp++)
		{
			// t \in T
			for(int t = 1; t < p.getTimePointSet().length; t++)
			{
				IloLinearNumExpr left = cplex.linearNumExpr();

				// i \in I
				for(int i = 0; i < p.getDemand().length; i++)
				{
					Request od = in.getRequestSet().get(i);

					// Input Z flow:
					// (item1)
					// o(i) == p
					if(p.getPortSet()[pp].equals(p.getOriginOfDemand()[i]))
					{
						//θi
						for (int k = 0; k < od.getNumberOfEmptyPath(); k++)
						{
							int j = od.getEmptyPathIndexes()[k];

							// <n,n'> ∈A'
							for (int nn = 0; nn < p.getTravelingArcsSet().length; nn++) {
								// p(n') == p
								// 1<= t(n')<= t
								if (in.getTravelingArcSet().get(nn).getDestinationPort().equals(p.getPortSet()[pp])
										&& in.getTravelingArcSet().get(nn).getDestinationTime() <= t
										&& in.getTravelingArcSet().get(nn).getDestinationTime() >= 1)
								{
									left.addTerm(p.getArcAndPath()[nn][j], zVar.get(i)[k]);
								}
							}
						}
					}


					// Input flow X
					// item2
					// d(i) == p
					if(p.getPortSet()[pp].equals(p.getDestinationOfDemand()[i]))
					{
						for(int k=0;k<od.getNumberOfLadenPath();k++)
						{
							int j = od.getLadenPathIndexes()[k];

							// <n,n'>∈A'
							for(int nn = 0; nn<p.getTravelingArcsSet().length; nn++)
							{
								// p(n‘)∈p
								// 1 <= t(n')<= t-sp
								if(in.getTravelingArcSet().get(nn).getDestinationPort().equals(p.getPortSet()[pp])
										&&in.getTravelingArcSet().get(nn).getDestinationTime()<=t-p.getTurnOverTime()[pp]
										&&in.getTravelingArcSet().get(nn).getDestinationTime()>=1)
								{
									left.addTerm(p.getArcAndPath()[nn][j], xVar.get(i)[k]);
								}
							}
						}
					}


					//Output  flow X
					// item3
					// o(i) == p
					if(p.getPortSet()[pp].equals(p.getOriginOfDemand()[i]))
					{
						// φi
						for(int k=0;k<od.getNumberOfLadenPath();k++)
						{
							int j = od.getLadenPathIndexes()[k];

							// <n.n'>∈A'
							for(int nn = 0; nn<p.getTravelingArcsSet().length; nn++)
							{
								//p(n) == p
								// t(n) <= t
								if(in.getTravelingArcSet().get(nn).getOriginPort().equals(p.getPortSet()[pp])
										&&in.getTravelingArcSet().get(nn).getOriginTime() <=t
										&&in.getTravelingArcSet().get(nn).getOriginTime() >=1)
								{
									left.addTerm(-p.getArcAndPath()[nn][j], xVar.get(i)[k]);
								}
							}
						}
					}


					// Output Flow Z
					// item4
					// θ
					for(int k=0;k<od.getNumberOfEmptyPath();k++)
					{
						int j = od.getEmptyPathIndexes()[k];

						// <n,n'>∈A'
						for(int nn = 0; nn<p.getTravelingArcsSet().length; nn++)
						{
							// p(n) == p
							// t(n) <= t
							if(in.getTravelingArcSet().get(nn).getOriginPort().equals(p.getPortSet()[pp])
									&& in.getTravelingArcSet().get(nn).getOriginTime()<=t
									&& in.getTravelingArcSet().get(nn).getOriginTime()>=1)
							{
								left.addTerm(-p.getArcAndPath()[nn][j], zVar.get(i)[k]);
							}
						}
					}
				}
				String ConstrName = "C3("+(pp+1)+")("+(t)+")";
				cplex.addGe(left, -p.getInitialEmptyContainer()[pp], ConstrName);
			}
		}
	}

	public void solveModel()
	{
		try
		{
			if (WhetherExportModel)
				exportModel();
			long startTime = System.currentTimeMillis();
			if (cplex.solve())
			{
				long endTime = System.currentTimeMillis();
				setObjVal(cplex.getObjValue());
				setSolveTime(endTime - startTime);
				setObjGap(cplex.getMIPRelativeGap());

				setvVarValue();
				setOperationCost();
			}
			else
			{
				System.out.println("No Solution");
			}
		}
		catch (IloException ex)
		{
			System.out.println("Concert Error: " + ex);
		}
	}
	private void setOperationCost() throws IloException {
		double operCost = 0;
		// r∈R
		for(int r = 0; r<p.getShippingRouteSet().length; r++)
		{
			// h∈H
			for(int h=0;h<p.getVesselSet().length;h++)
			{
				for (int w = 0; w < p.getVesselPathSet().length; w++) {
					operCost += p.getVesselOperationCost()[h]
							*p.getShipRouteAndVesselPath()[r][w]
							*p.getVesselTypeAndShipRoute()[h][r]
							*(int)(cplex.getValue(vVar[h][r]) + 0.5);
				}
			}
		}
		this.operationCost = operCost;
	}

	public void changeConstraints(double[] uValue) throws IloException {
		this.uValueDouble = uValue;
		/*
		Change Demand Equation Constraint(Constraint2)'s Right Coefficients
		 */
		//∀i∈I
		for(int i=0;i<p.getDemand().length;i++)
		{
		   C1[i].setBounds(p.getDemand()[i]+p.getMaximumDemandVariation()[i]* this.uValueDouble[i],
				   p.getDemand()[i]+p.getMaximumDemandVariation()[i]* this.uValueDouble[i]);
		}
	}

	public void setvVarValue() throws IloException {
		int[][] VValue = new	int[p.getVesselSet().length][p.getShippingRouteSet().length];
		for (int h = 0; h < p.getVesselSet().length; h++) {
			for (int r = 0; r < p.getShippingRouteSet().length; r++) {
				VValue[h][r] = (int)(cplex.getValue(vVar[h][r])+0.5);
			}
		}
		this.vVarValue = VValue;
	}

//	// An Unknown Error whiling get X
//	double[][] XValue;
//	private void setX() throws IloException {
//		this.XValue = new	double[p.getDemand().length][p.getPathSet().length];
//		for (int i = 0; i < p.getDemand().length; i++)
//		{
//			for (int j = 0; j < p.getPathSet().length; j++)
//			{
//				XValue[i][j] = cplex.getValue(xVar.get(i)[k]);
//			}
//		}
//	}
//	public double[][] getX()
//	{
//		return XValue;
//	}
}

File: Multi_routes_deploy/sources/model/SubDerModel.cpp
#include "SubDerModel.h"

namespace fleetdeployment {


	SubDerModel::SubDerModel(const InputData& in, const Parameter& p)
		: in_(in), p_(p)
		, uValue_(p_.GetDemand().size())
	{
		try {

			env_ = IloEnv();
			model_ = IloModel(env_);
			cplex_ = IloCplex(model_);


			cplex_.setOut(env_.getNullStream());
			cplex_.setParam(IloCplex::Param::WorkMem, MaxWorkMem);
			cplex_.setParam(IloCplex::Param::TimeLimit, MIPTimeLimit);
			cplex_.setParam(IloCplex::Param::MIP::Tolerances::MIPGap, MIPGapLimit);
			cplex_.setParam(IloCplex::Param::Threads, MaxThreads);

			// Create basic decision variables and add basic constraints
			SetDecisionVars();
			SetObjective();
			SetConstraints();
		}
		catch (const IloException& e) {
			std::cerr << "Error: " << e.getMessage() << std::endl;
		}
	}

#include <ilcplex/ilocplex.h>
#include <vector>
#include <string>

	// Assuming the rest of your class definition is as provided...

	void SubDerModel::SetDecisionVars() {
		// Initializing vVar_
		vVar_.resize(p_.GetVesselSet().size());
		for (size_t h = 0; h < p_.GetVesselSet().size(); ++h) {
			vVar_[h].resize(p_.GetVesselRouteSet().size());
			for (size_t r = 0; r < p_.GetVesselRouteSet().size(); ++r) {
				std::string varName = "v(" + std::to_string(h + 1) + "," + std::to_string(r + 1) + ")";
				vVar_[h][r] = IloIntVar(env_, 0, 1, varName.c_str());
			}
		}

		// Initializing xVar_, yVar_, zVar_, and gVar_
		xVar_.resize(p_.GetDemand().size());
		yVar_.resize(p_.GetDemand().size());
		zVar_.resize(p_.GetDemand().size());
		gVar_.resize(p_.GetDemand().size());

		for (size_t i = 0; i < p_.GetDemand().size(); ++i) {
			Request od = in_.GetRequests()[i];

			xVar_[i].resize(od.GetNumberOfLadenPath());
			yVar_[i].resize(od.GetNumberOfLadenPath());
			zVar_[i].resize(od.GetNumberOfEmptyPath());

			for (size_t k = 0; k < od.GetNumberOfLadenPath(); ++k) {
				std::string xVarName = "x(" + std::to_string(i + 1) + "," + std::to_string(k + 1) + ")";
				xVar_[i][k] = IloNumVar(env_, 0, IloInfinity, xVarName.c_str());

				std::string yVarName = "y(" + std::to_string(i + 1) + "," + std::to_string(k + 1) + ")";
				yVar_[i][k] = IloNumVar(env_, 0, IloInfinity, yVarName.c_str());
			}
			for (size_t k = 0; k < od.GetNumberOfEmptyPath(); ++k) {
				std::string zVarName = "z(" + std::to_string(i + 1) + "," + std::to_string(k + 1) + ")";
				zVar_[i][k] = IloNumVar(env_, 0, IloInfinity, zVarName.c_str());
			}

			std::string gVarName = "g(" + std::to_string(i + 1) + ")";
			gVar_[i] = IloNumVar(env_, 0, IloInfinity, gVarName.c_str());
		}
	}


	void SubDerModel::SetObjective() {
		IloEnv env = cplex_.getEnv();
		IloExpr Obj(env_);

		// item1: Operating Cost
		// Iterate over vessel routes (r∈R) and vessel paths (w∈Ω)
		for (int r = 0; r < p_.GetVesselRouteSet().size(); ++r) {
			for (int w = 0; w < p_.GetVesselPathSet().size(); ++w) {
				// Iterate over vessel set (h∈H)
				for (int h = 0; h < p_.GetVesselSet().size(); ++h) {
					// Add terms for vessel operating cost
					Obj += p_.GetVesselTypeAndShipRoute()[h][r]
						* p_.GetShipRouteAndVesselPath()[r][w]
						* p_.GetVesselOperationCost()[h]
						* vVar_[h][r];
				}
			}
		}

		// item2: Penalty Cost of unsatisfied Demand (c2i*Gi)
		// Iterate over all demands (i)
		for (int i = 0; i < p_.GetDemand().size(); ++i) {
			// Add terms for penalty cost
			Obj += p_.GetPenaltyCostForDemand()[i] * gVar_[i];

			Request od = in_.GetRequests()[i];
			// Laden and empty path costs
			// Iterate over laden paths (φi)
			for (int k = 0; k < od.GetNumberOfLadenPath(); ++k) {
				int j = od.GetLadenPathIndexes()[k];
				// Add terms for demurrage of self-owned and leased containers and rental cost on laden paths
				Obj += p_.GetLadenPathCost()[j] * xVar_[i][k];
				Obj += p_.GetLadenPathCost()[j] * yVar_[i][k];
				Obj += p_.GetRentalCost() * p_.GetTravelTimeOnPath()[j] * yVar_[i][k];
			}

			// Iterate over empty paths (θi)
			for (int k = 0; k < od.GetNumberOfEmptyPath(); ++k) {
				int j = od.GetEmptyPathIndexes()[k];
				// Add terms for demurrage of self-owned containers for empty path
				Obj += p_.GetEmptyPathCost()[j] * zVar_[i][k];
			}
		}

		// Add the objective function to the model
		model_.add(IloMinimize(env, Obj));
	}


	//(15)each ship route selected one vessel type
	void SubDerModel::SetConstraint0() {

		// r∈R (Iterate over vessel routes)
		for (int r = 0; r < p_.GetVesselRouteSet().size(); ++r) {
			IloExpr left(env_);

			// h∈H (Iterate over vessel set)
			for (int h = 0; h < p_.GetVesselSet().size(); ++h) {
				// Add terms to the expression
				left += p_.GetVesselTypeAndShipRoute()[h][r] * vVar_[h][r];
			}

			// Constraint name
			std::string constr_name = "C0(" + std::to_string(r + 1) + ")";
			// Add equality constraint to the model
			model_.add(IloRange(env_, 1, left, 1, constr_name.c_str()));
		}
	}


	//(21)demand equation
#include <ilcplex/ilocplex.h>
#include <vector>
#include <string>

// Inside the SubDerModel class...

	void SubDerModel::SetConstraint1() {
		C1_.resize(p_.GetDemand().size());
		IloEnv env = cplex_.getEnv();

		// ∀i∈I (Iterate over all demands)
		for (int i = 0; i < p_.GetDemand().size(); ++i) {
			IloExpr left(env_);
			Request od = in_.GetRequests()[i];

			// φ (Iterate over laden paths)
			for (int k = 0; k < od.GetNumberOfLadenPath(); ++k) {
				int j = od.GetLadenPathIndexes()[k];
				left += xVar_[i][k] + yVar_[i][k];
			}

			// Add gVar to the expression
			left += gVar_[i];

			// Construct constraint name
			std::string constr_name = "C1(" + std::to_string(i + 1) + ")";
			// Add equality constraint to the model
			C1_[i] = IloRange(env, left, p_.GetDemand()[i] + p_.GetMaximumDemandVariation()[i] * uValue_[i], constr_name.c_str());
			model_.add(C1_[i]);
		}
	}


	//(22) Vessel Capacity Constraint
	void SubDerModel::SetConstraint2() {
		IloEnv env = cplex_.getEnv();

		// ∀<n,n'>∈A' (Iterate over all travel arcs)
		for (int nn = 0; nn < p_.GetTravelArcsSet().size(); ++nn) {
			IloExpr left(env_);

			// i∈I (Iterate over all demands)
			for (int i = 0; i < p_.GetDemand().size(); ++i) {
				Request od = in_.GetRequests()[i];

				// φ (Iterate over laden paths)
				for (int k = 0; k < od.GetNumberOfLadenPath(); ++k) {
					int j = od.GetLadenPathIndexes()[k];
					left += p_.GetArcAndPath()[nn][j] * xVar_[i][k];
					left += p_.GetArcAndPath()[nn][j] * yVar_[i][k];
				}

				// θ (Iterate over empty paths)
				for (int k = 0; k < od.GetNumberOfEmptyPath(); ++k) {
					int j = od.GetEmptyPathIndexes()[k];
					left += p_.GetArcAndPath()[nn][j] * zVar_[i][k];
				}
			}

			// Vessel capacity
			// r∈R (Iterate over vessel routes)
			for (int r = 0; r < p_.GetVesselRouteSet().size(); ++r) {
				// ω∈Ω (Iterate over vessel paths)
				for (int w = 0; w < p_.GetVesselPathSet().size(); ++w) {
					// h∈H (Iterate over vessel set)
					for (int h = 0; h < p_.GetVesselSet().size(); ++h) {
						left -= p_.GetArcAndVesselPath()[nn][w]
							* p_.GetShipRouteAndVesselPath()[r][w]
							* p_.GetVesselTypeAndShipRoute()[h][r]
							* p_.GetVesselCapacity()[h]
							* vVar_[h][r];
					}
				}
			}

			// Construct constraint name
			std::string constr_name = "C3(" + std::to_string(nn + 1) + ")";
			// Add less-than-or-equal-to constraint to the model
			model_.add(IloRange(env, left, 0, constr_name.c_str()));
		}
	}

	// (24)Containers flow conservation
	void SubDerModel::SetConstraint3() {
		// p ∈ P (Iterate over ports)
		for (int pp = 0; pp < p_.GetPortSet().size(); ++pp) {
			// t ∈ T (Iterate over time points)
			for (int t = 1; t < p_.GetTimePointSet().size(); ++t) {
				IloExpr left(env_);

				// i ∈ I (Iterate over demands)
				for (int i = 0; i < p_.GetDemand().size(); ++i) {
					Request od = in_.GetRequests()[i];

					// Input Z flow:
					if (p_.GetPortSet()[pp] == p_.GetOriginOfDemand()[i]) {
						for (int k = 0; k < od.GetNumberOfEmptyPath(); ++k) {
							int j = od.GetEmptyPathIndexes()[k];
							for (int nn = 0; nn < p_.GetTravelArcsSet().size(); ++nn) {
								if (in_.GetTravelArcs()[nn].GetDestinationPort() == p_.GetPortSet()[pp] &&
									in_.GetTravelArcs()[nn].GetDestinationTime() <= t &&
									in_.GetTravelArcs()[nn].GetDestinationTime() >= 1) {
									left += p_.GetArcAndPath()[nn][j] * zVar_[i][k];
								}
							}
						}
					}

					// Input flow X
					if (p_.GetPortSet()[pp] == p_.GetDestinationOfDemand()[i]) {
						for (int k = 0; k < od.GetNumberOfLadenPath(); ++k) {
							int j = od.GetLadenPathIndexes()[k];
							for (int nn = 0; nn < p_.GetTravelArcsSet().size(); ++nn) {
								if (in_.GetTravelArcs()[nn].GetDestinationPort() == p_.GetPortSet()[pp] &&
									in_.GetTravelArcs()[nn].GetDestinationTime() <= t - p_.GetTurnOverTimeSet()[pp] &&
									in_.GetTravelArcs()[nn].GetDestinationTime() >= 1) {
									left += p_.GetArcAndPath()[nn][j] * xVar_[i][k];
								}
							}
						}
					}

					// Output flow X
					if (p_.GetPortSet()[pp] == p_.GetOriginOfDemand()[i]) {
						for (int k = 0; k < od.GetNumberOfLadenPath(); ++k) {
							int j = od.GetLadenPathIndexes()[k];
							for (int nn = 0; nn < p_.GetTravelArcsSet().size(); ++nn) {
								if (in_.GetTravelArcs()[nn].GetOriginPort() == p_.GetPortSet()[pp] &&
									in_.GetTravelArcs()[nn].GetOriginTime() <= t &&
									in_.GetTravelArcs()[nn].GetOriginTime() >= 1) {
									left -= p_.GetArcAndPath()[nn][j] * xVar_[i][k];
								}
							}
						}
					}

					// Output Flow Z
					for (int k = 0; k < od.GetNumberOfEmptyPath(); ++k) {
						int j = od.GetEmptyPathIndexes()[k];
						for (int nn = 0; nn < p_.GetTravelArcsSet().size(); ++nn) {
							if (in_.GetTravelArcs()[nn].GetOriginPort() == p_.GetPortSet()[pp] &&
								in_.GetTravelArcs()[nn].GetOriginTime() <= t &&
								in_.GetTravelArcs()[nn].GetOriginTime() >= 1) {
								left -= p_.GetArcAndPath()[nn][j] * zVar_[i][k];
							}
						}
					}
				}

				// Construct constraint name
				std::string constr_name = "C3(" + std::to_string(pp + 1) + "," + std::to_string(t) + ")";
				// Add greater-than-or-equal-to constraint to the model
				model_.add(IloRange(env_, -p_.GetInitialEmptyContainer()[pp], left, IloInfinity, constr_name.c_str()));
			}
		}
	}



	void SubDerModel::SolveModel() {
		try {
			// Uncomment the following line if you want to export the model
			// cplex_.exportModel("SDP1.lp");

			if (cplex_.solve()) {
				SetObjVal(cplex_.getObjValue());
				SetVVarValue();
				SetOperationCost();
			}
			else {
				std::cout << "No solution" << std::endl;
			}
		}
		catch (const IloException& ex) {
			std::cout << "Concert Error: " << ex.getMessage() << std::endl;
		}
	}

	// Assume SetObjVal, SetVVarValue, and SetOperationCost are correctly implemented


	void SubDerModel::changeConstraints(const std::vector<double>& uValue) {
		this->uValue_ = uValue; // Assuming uValue_ is a std::vector<double> member of SubDerModel

		// Change Demand Equation Constraint (Constraint1)'s Right Coefficients
		// ∀i∈I (Iterate over all demands)
		for (int i = 0; i < p_.GetDemand().size(); ++i) {
			double newBound = p_.GetDemand()[i] + p_.GetMaximumDemandVariation()[i] * this->uValue_[i];
			C1_[i].setBounds(newBound, newBound);
		}
	}


	void SubDerModel::SetVVarValue() {
		// Resize the 2D vector to match the dimensions of the vessel set and vessel route set
		vVarValue.resize(p_.GetVesselSet().size(), std::vector<int>(p_.GetVesselRouteSet().size()));

		// Iterate over vessel set and vessel route set
		for (int h = 0; h < p_.GetVesselSet().size(); ++h) {
			for (int r = 0; r < p_.GetVesselRouteSet().size(); ++r) {
				// Get the value of the decision variable and cast it to an integer
				vVarValue[h][r] = static_cast<int>(cplex_.getValue(vVar_[h][r]) + 0.5);
			}
		}
	}


	void SubDerModel::SetOperationCost() {
		double operCost = 0;
		// r∈R
		for (int r = 0; r < p_.GetVesselRouteSet().size(); r++)
		{
			// h∈H
			for (int h = 0; h < p_.GetVesselSet().size(); h++)
			{
				for (int w = 0; w < p_.GetVesselPathSet().size(); w++) {
					operCost += p_.GetVesselOperationCost()[h]
						* p_.GetShipRouteAndVesselPath()[r][w]
						* p_.GetVesselTypeAndShipRoute()[h][r]
						* (int)(cplex_.getValue(vVar_[h][r]) + 0.5);
				}
			}
		}
		operationCost = operCost;
	}




}